### ISSUE TITLE
Add a flag to print filenames as they're being compiled (even if they are not updated in output)

### ISSUE THREAD

**Issue by sirosen** [nanoc/nanoc#1504]
Here's my situation: I'm using nanoc with asciidoctor to build a static site from adoc files.
The trouble I have is specific to this combination, but might appear with other filters as well.

asciidoctor prints warnings to stderr, but without the context of which file was being processed, they aren't very useful.
If I could get nanoc to print the filenames being compiled (e.g. with `--verbose`), I would see a filename followed by a set of warnings.

The degree to which I can customize the asciidoctor behavior is limited. I can get it to show file basenames in its warnings, but not full paths. And its `-Werror`-like option only seems to work with command-line invocations.
There are some related open asciidoctor issues, and there's more that they could do on this too, to be sure. But my scenario lives at the intersection of these two tools, so it's equally reasonable for me to look to nanoc to solve it.

For now, I've added a custom filter which uses the asciidoctor gem (almost identical to the built-in asciidoctor filter) and checks for an env var, `NANOC_ADOC_DEBUG=1`. If it's set, it prints the item identifier every time it runs. It's a viable workaround, and I am very happy that nanoc makes it possible, but I'd be _even happier_ if nanoc had some flag I could set to ensure that all item names are printed.

**Comment by gpakosz** [comment]
I don't think you should reason in filenames.
What's given to Asciidoctor might not even be a file.

I'm not sure how but the stock `:asciidoctor` filter should imho capture errors and eventually print them along with item's identifier.

**Comment by sirosen** [comment]
> I don't think you should reason in filenames.
> What's given to Asciidoctor might not even be a file.

By this, I assume you're referring to the possibility of other data sources for items.
That's fair. _My_ case is using the filesystem, but the thing that matters to me is to be able to see a warning which a filter emitted on stderr and map that back to the item which produced the error.

If I could get item identifiers, my use-case would be satisfied -- because the identifiers will be filenames for me.

> I'm not sure how but the stock `:asciidoctor` filter should imho capture errors and eventually print them along with item's identifier.

Given that a filter can be calling out to an arbitrary library, it's possible that the lib in question is writing warnings directly to stdout or stderr. The only generic way I can think for nanoc to handle this, without enforcing that filters happen in a separate process, is to capture the standard streams, recording anything which the filter might write there, and then check if it's empty.

If the parallel build support ever merges, a sophisticated approach like that might be necessary in order to mux together the output from various filters.

But for today, just getting nanoc to emit the item identifiers would be sufficient and solve this problem.

**Comment by ddfreyne** [comment]
I think it’d be reasonable for Nanoc to print the item rep identifiers + names when `--verbose` is specified twice — what do you think?

> If the parallel build support ever merges

It won’t be soon, I’m afraid… I believe Nanoc needs some architectural rework for parallel compilation to be implemented in a solid, reliable way.

**Comment by ddfreyne** [comment]
Hmm — even without parallel compilation, the output is confusing. Take this as an example:

```
Loading site… done
Compiling site…
       start  /sitemap.xml.erb (rep name :default)
       start  /sitemap.xml.erb (rep name :latex)
       start  /favicon.ico (rep name :default)
       start  /favicon.ico (rep name :latex)
       start  /404.dmark (rep name :default)
       start  /release-notes.md (rep name :default)
      create  [0.00s]  output/favicon.ico
      create  [0.04s]  output/sitemap.xml
lib/helpers/release_notes.rb:19: warning: deprecated Object#=~ is called on Nokogiri::XML::Element; it always returns nil
       start  /404.dmark (rep name :default)
       start  /404.dmark (rep name :latex)
       start  /style-guide.dmark (rep name :latex)
       start  /style-guide.dmark (rep name :default)
      create  [0.39s]  output/release-notes/index.html
      create  [0.00s]  output/404.html
       start  /whats-new.dmark (rep name :latex)
```

Note this message:

```
lib/helpers/release_notes.rb:19: warning: deprecated Object#=~ is called on Nokogiri::XML::Element; it always returns nil
```

(which is probably something I should fix, but let’s ignore that for now)

This message appears after

```
      create  [0.00s]  output/favicon.ico
      create  [0.04s]  output/sitemap.xml
```

… which happens because Nanoc writes files async. So, it is not quite clear where messages are originating from.

**Comment by sirosen** [comment]
> I think it’d be reasonable for Nanoc to print the item rep identifiers + names when `--verbose` is specified twice — what do you think?

Yeah, `--verbose --verbose` is totally fine and reasonable in my book.
It was one of the options I tried when I was looking at whether not anything existed today -- before I even looked at documentation -- which suggests, anecdotally at least, that others might try it too.

I don't much care _how_ it's enabled. Mainly, I want to be able to turn this on in our CI builds.

> > If the parallel build support ever merges
> 
> It won’t be soon, I’m afraid… I believe Nanoc needs some architectural rework for parallel compilation to be implemented in a solid, reliable way.

Yeah, I figured it was difficult based on the fact that you've done a proof of concept a while back but it hasn't merged.

No pressure from me on this, certainly! We'd love to have parallel builds, of course, but cutting our build from 19 seconds to 10 seconds just isn't that important. 😄 

> This message appears after
> [...]
> which happens because Nanoc writes files async.

When I started emitting identifiers in my filter code, it always seemed to sync up correctly. Was I just getting lucky in my tests, or is there a step like "filter and write" which is being executed atomically?

I'm just curious if (1) my workaround is prone to failure (2) maybe there's something in how that's executed which helps show a way to do this?
